EIPS/eip-152.md

```diff
@@ -0,0 +1,232 @@
+- `m` - the message block vector
+- `t_0, t_1` - offset counters
+- `f` - the final block indicator flag
+
+```
+[4 bytes for rounds][64 bytes for h][128 bytes for m][8 bytes for t_0][8 bytes for t_1][1 byte for f]
+```
+
+`f` parameter is considered as `true` if it is `nonzero`. For example:
+- `00000000` is `false`
+- `00000001` is `true`
+- `10000000` is `true`
+- `11111111` is `true`
+
+The precompile should compute the `F` function as [specified in the RFC](https://github.com/keep-network/go-ethereum/pull/4) and return the updated state vector `h`.
+
+### Example Usage in Solidity
+
+The precompile can be wrapped easily in Solidity to provide a more development-friendly interface to `F`.
+
+```solidity
+function F(uint32 rounds, bytes32[2] memory h, bytes32[4] memory m, uint64[2] memory t, bool f) public view returns (bytes32[2] memory) {
+  bytes32[2] memory output;
+
+  bytes memory args = abi.encodePacked(rounds, h[0], h[1], m[0], m[1], m[2], m[3], t[0], t[1], f);
+
+  assembly {
+      if iszero(staticcall(not(0), 0x09, add(args, 32), 0xd5, output, 0x40)) {
+          revert(0, 0)
+      }
+  }
+
+  return output;
+}
+
+function callF() public view returns (bytes32[2] memory) {
+  bytes32[2] memory h;
+  h[0] = hex"6a09e627f3bcc909bb67ae8484caa73b3c6ef372fe94b82ba54ff53a5f1d36f2"; 
+  h[1] = hex"510e527fade682d19b05688c2b3e6c1f1f83d9abfb41bd6b5be0cd19137e2179";
+  
+  bytes32[4] memory m;
+  m[0] = hex"278400340e6b05c5752592c52c4f121292eafc51cc01a997b4aed13409298fad";
+  m[1] = hex"0d99ccc8f76453d9b3661e5c4d325d7751147db17489046d1682d50eefa4a1da";
+  m[2] = hex"0000000000000000000000000000000000000000000000000000000000000000";
+  m[3] = hex"0000000000000000000000000000000000000000000000000000000000000000";
+ 
+  uint64[2] memory t;
+  t[0] = 18446744073709551552;  // ffffffffffffffc0
+  t[1] = 18446744073709551615;	// ffffffffffffffff
+
+  bool f = true;
+
+  uint32 rounds = 12;
+
+  return F(rounds, h, m, t, f);
+}    
+```
+
+### Gas costs and benchmarks
+
+Each operation will cost `GFROUND * rounds` gas, where `GFROUND = 1`. Detailed benchmarks are presented in the benchmarks appendix section.
+
+## Rationale
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions  were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
+
+BLAKE2b is an excellent candidate for precompilation. It exhibits an extremely asymmetric efficiency. BLAKE2b is heavily optimized for modern 64-bit CPUs, specifically utilizing 24 and 63-bit rotations to allow parallelism through SIMD instructions and little-endian arithmetic. These characteristics provide exceptional speed on native CPUs: 3.08 cycles per byte, or 1 gibibyte per second on an Intel i5.
```

Comment:
What does "asymmetric efficiency" mean here?

Reply:
This was originally @tjade273, but I read it as "this is a heavily optimized operation, and it a great candidate for special-casing"

Reply:
"asymmetric"  in CS usually means one operation is way faster than the other, e.g. decompression is faster than compression. Perhaps the term is used with a specific meaning in the context of crypto/hashes.

I'm not just not sure it gives the right idea to the audience of this EIP.

Reply:
@pdyraga oh perhaps you want to consider this discussion here